// asana/dist/api/BudgetsApi.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BudgetsApi = void 0;
var _ApiClient = require("../ApiClient");
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); } /*
 * Asana
 * This is the interface for interacting with the [Asana Platform](https://developers.asana.com). Our API reference is generated from our [OpenAPI spec] (https://raw.githubusercontent.com/Asana/openapi/master/defs/asana_oas.yaml).
 *
 * OpenAPI spec version: 1.0
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 *
 * Swagger Codegen version: 3.0.54
 *
 * Do not edit the class manually.
 *
 */
var Collection = require('../utils/collection');

/**
* Budgets service.
* @module api/BudgetsApi
* @version 3.1.5
*/
var BudgetsApi = exports.BudgetsApi = /*#__PURE__*/function () {
  /**
  * Constructs a new BudgetsApi. 
  * @alias module:api/BudgetsApi
  * @class
  * @param {module:ApiClient} [apiClient] Optional API client implementation to use,
  * default to {@link module:ApiClient#instanc
  e} if unspecified.
  */
  function BudgetsApi(apiClient) {
    _classCallCheck(this, BudgetsApi);
    this.apiClient = apiClient || _ApiClient.ApiClient.instance;
  }

  /**
   * Create a budget
   * Creates a new budget.
   * @param {module:model/Object} body The budget to create.
   * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with an object containing data and HTTP response
   */
  return _createClass(BudgetsApi, [{
    key: "createBudgetWithHttpInfo",
    value: function createBudgetWithHttpInfo(body) {
      var postBody = body;
      // verify the required parameter 'body' is set
      if (body === undefined || body === null) {
        throw new Error("Missing the required parameter 'body' when calling createBudget");
      }
      var pathParams = {};
      var queryParams = {};
      var headerParams = {};
      var formParams = {};
      var authNames = ['personalAccessToken'];
      var contentTypes = ['application/json; charset=UTF-8'];
      var accepts = ['application/json; charset=UTF-8'];
      var returnType = 'Blob';
      return this.apiClient.callApi('/budgets', 'POST', pathParams, queryParams, headerParams, formParams, postBody, authNames, contentTypes, accepts, returnType);
    }

    /**
     * Create a budget
     * Creates a new budget.
     * @param {<&vendorExtensions.x-jsdoc-type>} body The budget to create.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/BudgetResponseData}
     */
  }, {
    key: "createBudget",
    value: function createBudget(body) {
      return this.createBudgetWithHttpInfo(body).then(function (response_and_data) {
        return response_and_data.data;
      });
    }

    /**
     * Delete a budget
     * A specific, existing budget can be deleted by making a DELETE request on the URL for that budget.  Returns an empty data record.
     * @param {String} budget_gid Globally unique identifier for the budget.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with an object containing data and HTTP response
     */
  }, {
    key: "deleteBudgetWithHttpInfo",
    value: function deleteBudgetWithHttpInfo(budget_gid) {
      var postBody = null;
      // verify the required parameter 'budget_gid' is set
      if (budget_gid === undefined || budget_gid === null) {
        throw new Error("Missing the required parameter 'budget_gid' when calling deleteBudget");
      }
      var pathParams = {
        'budget_gid': budget_gid
      };
      var queryParams = {};
      var headerParams = {};
      var formParams = {};
      var authNames = ['personalAccessToken'];
      var contentTypes = [];
      var accepts = ['application/json; charset=UTF-8'];
      var returnType = 'Blob';
      return this.apiClient.callApi('/budgets/{budget_gid}', 'DELETE', pathParams, queryParams, headerParams, formParams, postBody, authNames, contentTypes, accepts, returnType);
    }

    /**
     * Delete a budget
     * A specific, existing budget can be deleted by making a DELETE request on the URL for that budget.  Returns an empty data record.
     * @param {<&vendorExtensions.x-jsdoc-type>} budget_gid Globally unique identifier for the budget.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/EmptyResponseData}
     */
  }, {
    key: "deleteBudget",
    value: function deleteBudget(budget_gid) {
      return this.deleteBudgetWithHttpInfo(budget_gid).then(function (response_and_data) {
        return response_and_data.data;
      });
    }

    /**
     * Get a budget
     * Returns the complete budget record for a single budget.
     * @param {String} budget_gid Globally unique identifier for the budget.
     * @param {Object} opts Optional parameters
     * @param {Array.<module:model/String>} opts.opt_fields This endpoint returns a resource which excludes some properties by default. To include those optional properties, set this query parameter to a comma-separated list of the properties you wish to include.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with an object containing data and HTTP response
     */
  }, {
    key: "getBudgetWithHttpInfo",
    value: function getBudgetWithHttpInfo(budget_gid, opts) {
      opts = opts || {};
      var postBody = null;
      // verify the required parameter 'budget_gid' is set
      if (budget_gid === undefined || budget_gid === null) {
        throw new Error("Missing the required parameter 'budget_gid' when calling getBudget");
      }
      var pathParams = {
        'budget_gid': budget_gid
      };
      var queryParams = {};
      opts = opts || {};
      queryParams = opts;
      var headerParams = {};
      var formParams = {};
      var authNames = ['personalAccessToken'];
      var contentTypes = [];
      var accepts = ['application/json; charset=UTF-8'];
      var returnType = 'Blob';
      return this.apiClient.callApi('/budgets/{budget_gid}', 'GET', pathParams, queryParams, headerParams, formParams, postBody, authNames, contentTypes, accepts, returnType);
    }

    /**
     * Get a budget
     * Returns the complete budget record for a single budget.
     * @param {<&vendorExtensions.x-jsdoc-type>} budget_gid Globally unique identifier for the budget.
     * @param {Object} opts Optional parameters
     * @param {Array.<module:model/String>} opts.opt_fields This endpoint returns a resource which excludes some properties by default. To include those optional properties, set this query parameter to a comma-separated list of the properties you wish to include.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/BudgetResponseData}
     */
  }, {
    key: "getBudget",
    value: function getBudget(budget_gid, opts) {
      return this.getBudgetWithHttpInfo(budget_gid, opts).then(function (response_and_data) {
        return response_and_data.data;
      });
    }

    /**
     * Get all budgets
     * Gets all budgets for a given *parent*. This will at most return a list of size 1 for a given *parent*.
     * @param {String} parent Globally unique identifier for the budget&#x27;s parent object. This currently can only be a &#x60;project&#x60;.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with an object containing data and HTTP response
     */
  }, {
    key: "getBudgetsWithHttpInfo",
    value: function getBudgetsWithHttpInfo(parent) {
      var postBody = null;
      // verify the required parameter 'parent' is set
      if (parent === undefined || parent === null) {
        throw new Error("Missing the required parameter 'parent' when calling getBudgets");
      }
      var pathParams = {};
      var queryParams = {};
      queryParams['parent'] = parent;
      var headerParams = {};
      var formParams = {};
      var authNames = ['personalAccessToken'];
      var contentTypes = [];
      var accepts = ['application/json; charset=UTF-8'];
      var returnType = 'Blob';
      // Check if RETURN_COLLECTION is set and return a collection object if it is
      if (this.apiClient.RETURN_COLLECTION) {
        return Collection.fromApiClient(this.apiClient.callApi('/budgets', 'GET', pathParams, queryParams, headerParams, formParams, postBody, authNames, contentTypes, accepts, returnType), this.apiClient, {
          'path': '/budgets',
          'httpMethod': 'GET',
          'pathParams': pathParams,
          'queryParams': queryParams,
          'headerParams': headerParams,
          'formParams': formParams,
          'bodyParam': postBody,
          'authNames': authNames,
          'contentTypes': contentTypes,
          'accepts': accepts,
          'returnType': returnType
        });
      }
      return this.apiClient.callApi('/budgets', 'GET', pathParams, queryParams, headerParams, formParams, postBody, authNames, contentTypes, accepts, returnType);
    }

    /**
     * Get all budgets
     * Gets all budgets for a given *parent*. This will at most return a list of size 1 for a given *parent*.
     * @param {<&vendorExtensions.x-jsdoc-type>} parent Globally unique identifier for the budget&#x27;s parent object. This currently can only be a &#x60;project&#x60;.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/BudgetResponseArray}
     */
  }, {
    key: "getBudgets",
    value: function getBudgets(parent) {
      // Check if RETURN_COLLECTION is set and return a collection object if it is
      if (this.apiClient.RETURN_COLLECTION) {
        return this.getBudgetsWithHttpInfo(parent);
      }
      return this.getBudgetsWithHttpInfo(parent).then(function (response_and_data) {
        return response_and_data.data;
      });
    }

    /**
     * Update a budget
     * An existing budget can be updated by making a PUT request on the URL for that budget. Only the fields provided in the &#x60;data&#x60; block will be updated; any unspecified fields will remain unchanged.
     * @param {module:model/Object} body The budget to update.
     * @param {String} budget_gid Globally unique identifier for the budget.
     * @param {Object} opts Optional parameters
     * @param {Array.<module:model/String>} opts.opt_fields This endpoint returns a resource which excludes some properties by default. To include those optional properties, set this query parameter to a comma-separated list of the properties you wish to include.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with an object containing data and HTTP response
     */
  }, {
    key: "updateBudgetWithHttpInfo",
    value: function updateBudgetWithHttpInfo(body, budget_gid, opts) {
      opts = opts || {};
      var postBody = body;
      // verify the required parameter 'body' is set
      if (body === undefined || body === null) {
        throw new Error("Missing the required parameter 'body' when calling updateBudget");
      }
      // verify the required parameter 'budget_gid' is set
      if (budget_gid === undefined || budget_gid === null) {
        throw new Error("Missing the required parameter 'budget_gid' when calling updateBudget");
      }
      var pathParams = {
        'budget_gid': budget_gid
      };
      var queryParams = {};
      opts = opts || {};
      queryParams = opts;
      var headerParams = {};
      var formParams = {};
      var authNames = ['personalAccessToken'];
      var contentTypes = ['application/json; charset=UTF-8'];
      var accepts = ['application/json; charset=UTF-8'];
      var returnType = 'Blob';
      return this.apiClient.callApi('/budgets/{budget_gid}', 'PUT', pathParams, queryParams, headerParams, formParams, postBody, authNames, contentTypes, accepts, returnType);
    }

    /**
     * Update a budget
     * An existing budget can be updated by making a PUT request on the URL for that budget. Only the fields provided in the &#x60;data&#x60; block will be updated; any unspecified fields will remain unchanged.
     * @param {<&vendorExtensions.x-jsdoc-type>} body The budget to update.
     * @param {<&vendorExtensions.x-jsdoc-type>} budget_gid Globally unique identifier for the budget.
     * @param {Object} opts Optional parameters
     * @param {Array.<module:model/String>} opts.opt_fields This endpoint returns a resource which excludes some properties by default. To include those optional properties, set this query parameter to a comma-separated list of the properties you wish to include.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/BudgetResponseData}
     */
  }, {
    key: "updateBudget",
    value: function updateBudget(body, budget_gid, opts) {
      return this.updateBudgetWithHttpInfo(body, budget_gid, opts).then(function (response_and_data) {
        return response_and_data.data;
      });
    }
  }]);
}();
