// asana/dist/api/RatesApi.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RatesApi = void 0;
var _ApiClient = require("../ApiClient");
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); } /*
 * Asana
 * This is the interface for interacting with the [Asana Platform](https://developers.asana.com). Our API reference is generated from our [OpenAPI spec] (https://raw.githubusercontent.com/Asana/openapi/master/defs/asana_oas.yaml).
 *
 * OpenAPI spec version: 1.0
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 *
 * Swagger Codegen version: 3.0.54
 *
 * Do not edit the class manually.
 *
 */
var Collection = require('../utils/collection');

/**
* Rates service.
* @module api/RatesApi
* @version 3.1.5
*/
var RatesApi = exports.RatesApi = /*#__PURE__*/function () {
  /**
  * Constructs a new RatesApi. 
  * @alias module:api/RatesApi
  * @class
  * @param {module:ApiClient} [apiClient] Optional API client implementation to use,
  * default to {@link module:ApiClient#instanc
  e} if unspecified.
  */
  function RatesApi(apiClient) {
    _classCallCheck(this, RatesApi);
    this.apiClient = apiClient || _ApiClient.ApiClient.instance;
  }

  /**
   * Create a rate
   * Creates a new rate for a &#x60;parent&#x60; + &#x60;resource&#x60; combination.
   * @param {module:model/Object} body The rate to create.
   * @param {Object} opts Optional parameters
   * @param {Array.<module:model/String>} opts.opt_fields This endpoint returns a resource which excludes some properties by default. To include those optional properties, set this query parameter to a comma-separated list of the properties you wish to include.
   * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with an object containing data and HTTP response
   */
  return _createClass(RatesApi, [{
    key: "createRateWithHttpInfo",
    value: function createRateWithHttpInfo(body, opts) {
      opts = opts || {};
      var postBody = body;
      // verify the required parameter 'body' is set
      if (body === undefined || body === null) {
        throw new Error("Missing the required parameter 'body' when calling createRate");
      }
      var pathParams = {};
      var queryParams = {};
      opts = opts || {};
      queryParams = opts;
      var headerParams = {};
      var formParams = {};
      var authNames = ['personalAccessToken'];
      var contentTypes = ['application/json; charset=UTF-8'];
      var accepts = ['application/json; charset=UTF-8'];
      var returnType = 'Blob';
      return this.apiClient.callApi('/rates', 'POST', pathParams, queryParams, headerParams, formParams, postBody, authNames, contentTypes, accepts, returnType);
    }

    /**
     * Create a rate
     * Creates a new rate for a &#x60;parent&#x60; + &#x60;resource&#x60; combination.
     * @param {<&vendorExtensions.x-jsdoc-type>} body The rate to create.
     * @param {Object} opts Optional parameters
     * @param {Array.<module:model/String>} opts.opt_fields This endpoint returns a resource which excludes some properties by default. To include those optional properties, set this query parameter to a comma-separated list of the properties you wish to include.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/RateResponseData}
     */
  }, {
    key: "createRate",
    value: function createRate(body, opts) {
      return this.createRateWithHttpInfo(body, opts).then(function (response_and_data) {
        return response_and_data.data;
      });
    }

    /**
     * Delete a rate
     * Deletes a rate.
     * @param {String} rate_gid Globally unique identifier for the rate.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with an object containing data and HTTP response
     */
  }, {
    key: "deleteRateWithHttpInfo",
    value: function deleteRateWithHttpInfo(rate_gid) {
      var postBody = null;
      // verify the required parameter 'rate_gid' is set
      if (rate_gid === undefined || rate_gid === null) {
        throw new Error("Missing the required parameter 'rate_gid' when calling deleteRate");
      }
      var pathParams = {
        'rate_gid': rate_gid
      };
      var queryParams = {};
      var headerParams = {};
      var formParams = {};
      var authNames = ['personalAccessToken'];
      var contentTypes = [];
      var accepts = ['application/json; charset=UTF-8'];
      var returnType = 'Blob';
      return this.apiClient.callApi('/rates/{rate_gid}', 'DELETE', pathParams, queryParams, headerParams, formParams, postBody, authNames, contentTypes, accepts, returnType);
    }

    /**
     * Delete a rate
     * Deletes a rate.
     * @param {<&vendorExtensions.x-jsdoc-type>} rate_gid Globally unique identifier for the rate.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/EmptyResponseData}
     */
  }, {
    key: "deleteRate",
    value: function deleteRate(rate_gid) {
      return this.deleteRateWithHttpInfo(rate_gid).then(function (response_and_data) {
        return response_and_data.data;
      });
    }

    /**
     * Get a rate
     * Returns the complete rate record for a single rate.
     * @param {String} rate_gid Globally unique identifier for the rate.
     * @param {Object} opts Optional parameters
     * @param {Array.<module:model/String>} opts.opt_fields This endpoint returns a resource which excludes some properties by default. To include those optional properties, set this query parameter to a comma-separated list of the properties you wish to include.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with an object containing data and HTTP response
     */
  }, {
    key: "getRateWithHttpInfo",
    value: function getRateWithHttpInfo(rate_gid, opts) {
      opts = opts || {};
      var postBody = null;
      // verify the required parameter 'rate_gid' is set
      if (rate_gid === undefined || rate_gid === null) {
        throw new Error("Missing the required parameter 'rate_gid' when calling getRate");
      }
      var pathParams = {
        'rate_gid': rate_gid
      };
      var queryParams = {};
      opts = opts || {};
      queryParams = opts;
      var headerParams = {};
      var formParams = {};
      var authNames = ['personalAccessToken'];
      var contentTypes = [];
      var accepts = ['application/json; charset=UTF-8'];
      var returnType = 'Blob';
      return this.apiClient.callApi('/rates/{rate_gid}', 'GET', pathParams, queryParams, headerParams, formParams, postBody, authNames, contentTypes, accepts, returnType);
    }

    /**
     * Get a rate
     * Returns the complete rate record for a single rate.
     * @param {<&vendorExtensions.x-jsdoc-type>} rate_gid Globally unique identifier for the rate.
     * @param {Object} opts Optional parameters
     * @param {Array.<module:model/String>} opts.opt_fields This endpoint returns a resource which excludes some properties by default. To include those optional properties, set this query parameter to a comma-separated list of the properties you wish to include.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/RateResponseData}
     */
  }, {
    key: "getRate",
    value: function getRate(rate_gid, opts) {
      return this.getRateWithHttpInfo(rate_gid, opts).then(function (response_and_data) {
        return response_and_data.data;
      });
    }

    /**
     * Get multiple rates
     * Returns a list of &#x60;rate&#x60; records. The possible types for &#x60;parent&#x60; in this request are &#x60;project&#x60;. An additional &#x60;resource&#x60; (&#x60;user&#x60; GID or &#x60;placeholder&#x60; GID) can be passed in to filter to a specific rate.  Modifying placeholder rates is only available for Enterprise and Enterprise+ users.
     * @param {Object} opts Optional parameters
     * @param {String} opts.parent Globally unique identifier for &#x60;project&#x60;.
     * @param {String} opts.resource Globally unique identifier for &#x60;user&#x60; or &#x60;placeholder&#x60;.
     * @param {Number} opts.limit Results per page. The number of objects to return per page. The value must be between 1 and 100.
     * @param {String} opts.offset Offset token. An offset to the next page returned by the API. A pagination request will return an offset token, which can be used as an input parameter to the next request. If an offset is not passed in, the API will return the first page of results. *Note: You can only pass in an offset that was returned to you via a previously paginated request.*
     * @param {Array.<module:model/String>} opts.opt_fields This endpoint returns a resource which excludes some properties by default. To include those optional properties, set this query parameter to a comma-separated list of the properties you wish to include.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with an object containing data and HTTP response
     */
  }, {
    key: "getRatesWithHttpInfo",
    value: function getRatesWithHttpInfo(opts) {
      opts = opts || {};
      var postBody = null;
      var pathParams = {};
      var queryParams = {};
      opts = opts || {};
      queryParams = opts;
      var headerParams = {};
      var formParams = {};
      var authNames = ['personalAccessToken'];
      var contentTypes = [];
      var accepts = ['application/json; charset=UTF-8'];
      var returnType = 'Blob';
      // Check if RETURN_COLLECTION is set and return a collection object if it is
      if (this.apiClient.RETURN_COLLECTION) {
        return Collection.fromApiClient(this.apiClient.callApi('/rates', 'GET', pathParams, queryParams, headerParams, formParams, postBody, authNames, contentTypes, accepts, returnType), this.apiClient, {
          'path': '/rates',
          'httpMethod': 'GET',
          'pathParams': pathParams,
          'queryParams': queryParams,
          'headerParams': headerParams,
          'formParams': formParams,
          'bodyParam': postBody,
          'authNames': authNames,
          'contentTypes': contentTypes,
          'accepts': accepts,
          'returnType': returnType
        });
      }
      return this.apiClient.callApi('/rates', 'GET', pathParams, queryParams, headerParams, formParams, postBody, authNames, contentTypes, accepts, returnType);
    }

    /**
     * Get multiple rates
     * Returns a list of &#x60;rate&#x60; records. The possible types for &#x60;parent&#x60; in this request are &#x60;project&#x60;. An additional &#x60;resource&#x60; (&#x60;user&#x60; GID or &#x60;placeholder&#x60; GID) can be passed in to filter to a specific rate.  Modifying placeholder rates is only available for Enterprise and Enterprise+ users.
     * @param {Object} opts Optional parameters
     * @param {String} opts.parent Globally unique identifier for &#x60;project&#x60;.
     * @param {String} opts.resource Globally unique identifier for &#x60;user&#x60; or &#x60;placeholder&#x60;.
     * @param {Number} opts.limit Results per page. The number of objects to return per page. The value must be between 1 and 100.
     * @param {String} opts.offset Offset token. An offset to the next page returned by the API. A pagination request will return an offset token, which can be used as an input parameter to the next request. If an offset is not passed in, the API will return the first page of results. *Note: You can only pass in an offset that was returned to you via a previously paginated request.*
     * @param {Array.<module:model/String>} opts.opt_fields This endpoint returns a resource which excludes some properties by default. To include those optional properties, set this query parameter to a comma-separated list of the properties you wish to include.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/RateOrPlaceholderCompactArray}
     */
  }, {
    key: "getRates",
    value: function getRates(opts) {
      // Check if RETURN_COLLECTION is set and return a collection object if it is
      if (this.apiClient.RETURN_COLLECTION) {
        return this.getRatesWithHttpInfo(opts);
      }
      return this.getRatesWithHttpInfo(opts).then(function (response_and_data) {
        return response_and_data.data;
      });
    }

    /**
     * Update a rate
     * An existing rate can be updated by making a PUT request on the URL for that rate. Only the fields provided in the &#x60;data&#x60; block will be updated; any unspecified fields will remain unchanged. (note that at this time, the only field that can be updated is the &#x60;rate&#x60; field.)  Returns the complete updated rate record.
     * @param {module:model/Object} body The updated fields for the rate.
     * @param {String} rate_gid Globally unique identifier for the rate.
     * @param {Object} opts Optional parameters
     * @param {Array.<module:model/String>} opts.opt_fields This endpoint returns a resource which excludes some properties by default. To include those optional properties, set this query parameter to a comma-separated list of the properties you wish to include.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with an object containing data and HTTP response
     */
  }, {
    key: "updateRateWithHttpInfo",
    value: function updateRateWithHttpInfo(body, rate_gid, opts) {
      opts = opts || {};
      var postBody = body;
      // verify the required parameter 'body' is set
      if (body === undefined || body === null) {
        throw new Error("Missing the required parameter 'body' when calling updateRate");
      }
      // verify the required parameter 'rate_gid' is set
      if (rate_gid === undefined || rate_gid === null) {
        throw new Error("Missing the required parameter 'rate_gid' when calling updateRate");
      }
      var pathParams = {
        'rate_gid': rate_gid
      };
      var queryParams = {};
      opts = opts || {};
      queryParams = opts;
      var headerParams = {};
      var formParams = {};
      var authNames = ['personalAccessToken'];
      var contentTypes = ['application/json; charset=UTF-8'];
      var accepts = ['application/json; charset=UTF-8'];
      var returnType = 'Blob';
      return this.apiClient.callApi('/rates/{rate_gid}', 'PUT', pathParams, queryParams, headerParams, formParams, postBody, authNames, contentTypes, accepts, returnType);
    }

    /**
     * Update a rate
     * An existing rate can be updated by making a PUT request on the URL for that rate. Only the fields provided in the &#x60;data&#x60; block will be updated; any unspecified fields will remain unchanged. (note that at this time, the only field that can be updated is the &#x60;rate&#x60; field.)  Returns the complete updated rate record.
     * @param {<&vendorExtensions.x-jsdoc-type>} body The updated fields for the rate.
     * @param {<&vendorExtensions.x-jsdoc-type>} rate_gid Globally unique identifier for the rate.
     * @param {Object} opts Optional parameters
     * @param {Array.<module:model/String>} opts.opt_fields This endpoint returns a resource which excludes some properties by default. To include those optional properties, set this query parameter to a comma-separated list of the properties you wish to include.
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/RateResponseData}
     */
  }, {
    key: "updateRate",
    value: function updateRate(body, rate_gid, opts) {
      return this.updateRateWithHttpInfo(body, rate_gid, opts).then(function (response_and_data) {
        return response_and_data.data;
      });
    }
  }]);
}();
